File: 011-020/011.js
// grid is N * n
const N = 20;
// d個抽出する
const D = 4;
let max_index = N - D;

// 011.data.txtを読み込む
const path = require('path');
const fs = require('fs');
let input_raw = fs.readFileSync(path.join(__dirname, './011.data.txt'), 'utf-8');
let str_grid = input_raw.trim();

// gridの配列化
let grid_nums = [];
str_grid.split(/\r\n|\r|\n/).forEach((row) => {
  grid_nums.push(
    row.split(" ").map(function (value, index, arr) {
      return parseInt(value, 10)
    })
  );
});

let calc_result = 0;
for (let h = 0; h < N; h++) {
  for (let w = 0; w < N; w++) {
    //横の積
    if (w <= max_index) {
      let tmp_sum = 1
      for (let i = 0; i < D; i++) {
        tmp_sum *= grid_nums[h][w + i];
      }
      // tmp_sumが大きければ答えの候補としてメモリに格納する
      tmp_sum < calc_result || (calc_result = tmp_sum)
    }

    // 縦の積
    // 縦の座標が、自分含めて下からd個抽出できないならループを進める
    if (h > max_index) continue;
    let tmp_sum = 1
    for (let i = 0; i < D; i++) {
      tmp_sum *= grid_nums[h + i][w];
    }
    // tmp_sumが大きければ答えの候補としてメモリに格納する
    tmp_sum < calc_result || (calc_result = tmp_sum)

    // 斜めの積
    // まず左上から右下に足していく
    if (w <= max_index) {
      let tmp_sum = 1
      for (var i = 0; i < D; i++) {
        tmp_sum *= grid_nums[h + i][w + i];
      }
      // tmp_sumが大きければ答えの候補としてメモリに格納する
      tmp_sum < calc_result || (calc_result = tmp_sum)
    }
    // まず右上から左下に足していく
    if (w >= D - 1) {
      let tmp_sum = 1
      for (var i = 0; i < D; i++) {
        tmp_sum *= grid_nums[h + i][w - i];
      }
      // tmp_sumが大きければ答えの候補としてメモリに格納する
      tmp_sum < calc_result || (calc_result = tmp_sum)
    }
  }
}

console.log(calc_result);

File: 011-020/011.py
import os

# 問題を読み込む
grid_nums = None

filename = '011.data.txt'
path = os.path.join(os.path.dirname(__file__), filename)
with open(path) as f:
    grid_nums = [list(map(int, s.split(' '))) for s in f.readlines()]

d = 4
grid_size = len(grid_nums)
max_index = grid_size - d
calc_result = 0

for h in range(grid_size):
    for w in range(grid_size):
        # 横の積
        if w <= max_index:
            tmp_sum = 1
            for a in range(d):
                tmp_sum *= grid_nums[h][w + a]
            if calc_result < tmp_sum:
                calc_result = tmp_sum

        # 縦の積
        if h > max_index:
            continue
        tmp_sum = 1
        for a in range(d):
            tmp_sum *= grid_nums[h + a][w]
        if calc_result < tmp_sum:
            calc_result = tmp_sum

        # 斜めの積 斜めの場合、上から下に足すので縦の条件は同じ
        # 斜めで、左上から右下に評価していく時
        if w <= max_index:
            tmp_sum = 1
            for a in range(d):
                tmp_sum *= grid_nums[h + a][w + a]
            if calc_result < tmp_sum:
                calc_result = tmp_sum
        # 斜めで、右上から左下に評価していく時
        if w < d - 1:
            continue
        tmp_sum = 1
        for a in range(d):
            tmp_sum *= grid_nums[h + a][w - a]
        if calc_result < tmp_sum:
            calc_result = tmp_sum

print(calc_result)
